### PR TITLE
mgmt: mcumgr: lib: cmd: dev: Add device information handler

### PR DESCRIPTION
**Pull request by nordicjm**
PR for adding a device information mcumgr group. Functionality is similar to the linux `uname` command where a format string can be provided to return certain information including board name, processor type, architecture, kernel name/version, OS name. It can be extended by other code using the revamped mcumgr callback system to add additional output/processing or to replace the OS name.

    The device information handler can be used to retrieve information about
    the configuration of the configured device such as board name, board
    revision, firmware version and build date.

Example output, no parameters:
`Zephyr`
'a' parameter, with default Bluetooth enabled settings:
`Zephyr Zephyr zephyr-v3.0.0-2569-g7f0654f969f7 3.0.99 arm unknown bt510 Zephyr`

- [X] Base implementation
- [X] Build date/time support (with note about invalidating reproducible builds, disabled by default)
- [X] Tests (check right response is received, check if memory error is received if buffer size is too small, check overriding OS name)
- [X] Release notes update
- [X] Documentation

This requires https://github.com/zephyrproject-rtos/zephyr/pull/50578 as a base and has pulled those commits into this review (first 7 commits of this PR can be ignored)

### CODE REVIEW

**Review by de-nordic — CHANGES_REQUESTED** [review]
I think that this can be addition to OS group, there is no need to create additional group for that.
Consider also making the command more complex, like for example BSD `uname`, where request could be filled with optional "format" string that would be assigned letters representing requested info, for example
```
format="bfa";
```
would give response:
```
b="<CONFIG_BOARD>"
f="0.1.2-something"
unsupported="a"
```

**Comment by lairdjm** [comment]
@de-nordic Moved to OS and uname-style output, still some parts to implement. Thoughts on this? Not sure if the processor response is already available or if it would become an #ifdef mess with many types of processors and processor families. Is there a way to currently get the overall package that you know of e.g. "Zephyr" for if purely using zephyr or "nRF connect SDK" if using that, etc.?

File: subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig
Comment (lairdjm) [inline subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig:121]: Fix

File: subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c
Comment (lairdjm) [inline subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:10]: Fix newline

**Comment by de-nordic** [comment]
> @de-nordic Moved to OS and uname-style output, still some parts to implement. Thoughts on this? Not sure if the processor response is already available or if it would become an #ifdef mess with many types of processors and processor families. Is there a way to currently get the overall package that you know of e.g. "Zephyr" for if purely using zephyr or "nRF connect SDK" if using that, etc.?

I do not think that we have a way to get processor name for now.
During compilation we do not know processor name only the arch, maybe processor name could be pulled out of DTS somehow.
We can probably postpone the processor part and complete all other "letters".

File: subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig:124]: Invent your own name for this. Uname is Unix wide command and we have Zephyr here. Make some "os info", "zephyr info" or something. We can use similar option letters as "uname" but command should be ours.

Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/os_mgmt/Kconfig]: Fixed.

**Comment by lairdjm** [comment]
Added a custom function to allow additional options in application, and changed OS to be it's own (weak) function to allow overriding by vendors

**Comment by de-nordic** [comment]
@nordicjm Should we bring this back to live?

**Comment by nordicjm** [comment]
Would be a good feature to add, yes.

**Comment by de-nordic** [comment]
> Would be a good feature to add, yes.

Can we re-open this and set target to 3.3?

File: subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c
Comment (nordicjm) [inline subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c:471]: Needs addressing

File: tests/subsys/mgmt/mcumgr/os_mgmt_info/src/smp_test_util.c
Comment (de-nordic) [inline tests/subsys/mgmt/mcumgr/os_mgmt_info/src/smp_test_util.c:24]: This is internal function of SMP(rotocol) processing that should have never been exported, but mgmt interface is a mess so it was. It is even named wrongly.
This test implements client so it can spare some additional line to implement the client side of header encoding.

Comment (nordicjm) [reply tests/subsys/mgmt/mcumgr/os_mgmt_info/src/smp_test_util.c]: Fixed

**Comment by nordicjm** [comment]
Rebased

**Review by de-nordic — CHANGES_REQUESTED** [review]
One comment regarding Kconfig names.

**Comment by nordicjm** [comment]
@de-nordic Changes made, left some of the text as-is to prevent duplicating strings

**Review by de-nordic — APPROVED** [review]
I do not like the `(prior_output == true ... ` things in `sprintf` as I think that saving on strings will not be so significant with branching overhead.
Having said that I do not find it a deal breaker, as this is something we can argue on later and improve (or not).

Looks OK, Thanks.

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri can you merge this please?

Needs a second review, I'll take a look.

**Review by fabiobaltieri — APPROVED** [review]

File: subsys/mgmt/mcumgr/grp/os_mgmt/include/os_mgmt_processor.h
Comment (nordicjm) [reply subsys/mgmt/mcumgr/grp/os_mgmt/include/os_mgmt_processor.h:19]: Indeed it is horrible to look at and update, indentation would probably be a good idea, I'll give it a try and see if CI throws an error about it

Comment (fabiobaltieri) [reply subsys/mgmt/mcumgr/grp/os_mgmt/include/os_mgmt_processor.h:19]: Wondering if it's even worth reporting it in such details, maybe you could just report whatever is in `cpu@0` compatible.

**Pull request merged**
